src/main/kotlin/org/rust/lang/core/macros/MacroExpansionManager.kt

```diff
@@ -218,32 +224,114 @@ class MacroExpansionManagerImpl(
-    val expansionsDir = baseProjectDir.resolve("expansions")
+    val baseProjectDir = getBaseMacroDir()
         .also { it.createDirectories() }
     return Dirs(
-        baseProjectDir,
-        baseProjectDir.resolve("data.dat"),
-        LocalFileSystem.getInstance().refreshAndFindFileByIoFile(expansionsDir.toFile())!!
+        baseProjectDir.resolve("$projectDirName.dat"),
+        projectDirName
     )
 }
 
 private data class Dirs(
-    val baseProjectDir: Path,
     val dataFile: Path,
-    val expansionsDirVi: VirtualFile
-)
+    val projectDirName: String
+) {
+    // Path in the MacroExpansionVFS
+    val expansionDirPath: String get() = "/$MACRO_EXPANSION_VFS_ROOT/$projectDirName"
+}
 
 private class MacroExpansionServiceImplInner(
     private val project: Project,
     val dirs: Dirs,
-    private val storage: ExpandedMacroStorage
+    private val storage: ExpandedMacroStorage,
+    var expansionsDirVi: VirtualFile?
 ) {
+    companion object {
+        fun build(project: Project, dirs: Dirs): () -> MacroExpansionServiceImplInner {
+            val dataFile = dirs.dataFile
+            loadProjectDirs()
+            val loaded = load(dataFile)
+
+            val vfs = MacroExpansionFileSystem.getInstance()
+            return if (loaded != null) {
+                val (serStorage, fs) = loaded
+                vfs.setDirectory(dirs.expansionDirPath, fs)
+                val dir = vfs.findFileByPath(dirs.expansionDirPath) // Nullable!
+                fun(): MacroExpansionServiceImplInner {
+                    val storage = serStorage.deserializeInReadAction(project)
+                    return MacroExpansionServiceImplInner(project, dirs, storage, dir)
+                }
+            } else {
+                MACRO_LOG.debug("Using fresh ExpandedMacroStorage")
+                vfs.createDirectoryIfNotExistsOrDummy(dirs.expansionDirPath)
+                val dir = vfs.findFileByPath(dirs.expansionDirPath) // Nullable!
+                val mes = MacroExpansionServiceImplInner(project, dirs, ExpandedMacroStorage(project), dir)
+                fun(): MacroExpansionServiceImplInner = mes
+            }
+        }
+
+        private fun load(dataFile: Path): Pair<SerializedExpandedMacroStorage, MacroExpansionFileSystem.FSItem.FSDir>? {
+            return try {
+                dataFile.newInflaterDataInputStream().use { data ->
+                    val sems = SerializedExpandedMacroStorage.load(data) ?: return null
+                    val fs = MacroExpansionFileSystem.readFSItem(data, null) as? MacroExpansionFileSystem.FSItem.FSDir ?: return null
+                    sems to fs
+                }
+            } catch (e: java.nio.file.NoSuchFileException) {
+                null
+            } catch (e: Exception) {
+                MACRO_LOG.warn(e)
+                null
+            }
+        }
+
+        /** See [MacroExpansionFileSystem] docs for explanation of what happens here */
+        private fun loadProjectDirs() {
```

Comment:
Hm, what if we call this method from two different projects at the same time?
It seems we need some synchronization here to avoid `File already exists` exception